peps/pep-0788.rst

```diff
@@ -0,0 +1,619 @@
+
+    Decrement the reference count of the interpreter. This function mainly
+    exists for completeness, and should rarely be used; nearly all references
+    returned by :c:func:`PyInterpreterState_Hold` should be released by
+    :c:func:`PyThreadState_Ensure`.
+
+    This function cannot fail, other than with a fatal error. The caller must
+    have an :term:`attached thread state` for *interp*.
+
+
+Daemon and non-daemon threads
+-----------------------------
+
+This PEP introduces the concept of non-daemon thread states. By default, all
+threads created without the :mod:`threading` module will hang when trying to
+attach a thread state for a finalizing interpreter (in fact, daemon threads
+that *are* created with the :mod:`threading` module will hang in the same
+way). This generally happens when a thread calls :c:func:`PyEval_RestoreThread`
+or in between bytecode instructions, based on :func:`sys.setswitchinterval`.
+
+A new, internal field will be added to the ``PyThreadState`` structure that
+determines if the thread is daemon. If the thread is daemon, then it will
+hang during attachment as usual, but if it's not, then the interpreter will
+let the thread attach and continue execution. On with-GIL builds, this again
+means handing off the GIL to the thread. During finalization, the interpreter
+will wait until all non-daemon threads call :c:func:`PyThreadState_Delete`.
+
+For backwards compatibility, all thread states created by existing APIs will
+remain daemon by default.
+
+.. c:function:: int PyThreadState_SetDaemon(int is_daemon)
+
+    Set the :term:`attached thread state` as non-daemon or daemon. The attached
+    thread state must not be the main thread for the interpreter. All thread
+    states created without :c:func:`PyThreadState_Ensure` are daemon by
+    default.
+
+    If the thread state is non-daemon, then the current interpreter will wait
+    for this thread to finish before shutting down. See also
+    :attr:`threading.Thread.daemon`.
+
+    Return zero on success, non-zero *without* an exception set on failure.
+    Failure generally means that threads have already finalized for the
+    current interpreter.
+
+Ensuring and releasing thread states
+------------------------------------
+
+This proposal includes two new high-level threading APIs that intend to
+replace :c:func:`PyGILState_Ensure` and :c:func:`PyGILState_Release`.
+
+.. c:function:: int PyThreadState_Ensure(PyInterpreterState *interp)
+
+    Ensure that the thread has an :term:`attached thread state` for *interp*, and
+    thus can safely invoke that interpreter. It is OK to call this function if
+    the thread already has an attached thread state, as long as there is a
+    subsequent call to :c:func:`PyThreadState_Release` that matches this one.
+
+    This function always steals a reference to *interp*; as in, the
+    interpreter's reference count is decremented by one. As such, *interp*
```

Comment:
I would prefer to avoid "steal" term here:

```suggestion
    The interpreter's *interp* reference count is decremented by one. As such, *interp*
```